File: jvm/examples/src/main/java/pt/isel/pc/demos/li51d/synchronizers/BlockingQueue.java
package pt.isel.pc.demos.li51d.synchronizers;

import pt.isel.pc.utils.NodeLinkedList;
import pt.isel.pc.utils.Timeouts;

import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class BlockingQueue<E> {

    static class SendRequest<E> {
        boolean isDone = false;
        final Condition condition;
        final E elementToSend;

        SendRequest(Condition condition, E elementToSend) {
            this.condition = condition;
            this.elementToSend = elementToSend;
        }
    }

    static class ReceiveRequest<E> {
        boolean isDone = false;
        final Condition condition;
        E receivedElement = null;

        ReceiveRequest(Condition condition) {
            this.condition = condition;
        }
    }

    private final Lock monitor = new ReentrantLock();
    private final NodeLinkedList<SendRequest<E>> sendQueue = new NodeLinkedList<>();
    private final NodeLinkedList<ReceiveRequest<E>> receiveQueue = new NodeLinkedList<>();

    public boolean enqueue(E element, long timeout, TimeUnit timeUnit) throws InterruptedException {
        monitor.lock();
        try {

            // fast path
            if (receiveQueue.isNotEmpty()) {
                NodeLinkedList.Node<ReceiveRequest<E>> receiver = receiveQueue.pull();
                receiver.value.receivedElement = element;
                receiver.value.isDone = true;
                receiver.value.condition.signal();
                return true;
            }
            if (Timeouts.noWait(timeout)) {
                return false;
            }

            // wait path
            long deadline = Timeouts.start(timeout, timeUnit);
            long remaining = Timeouts.remaining(deadline);
            NodeLinkedList.Node<SendRequest<E>> node =
                    sendQueue.enqueue(new SendRequest<>(monitor.newCondition(), element));
            while (true) {
                try {
                    node.value.condition.await(remaining, TimeUnit.MILLISECONDS);
                } catch (InterruptedException e) {
                    if (node.value.isDone) {
                        Thread.currentThread().interrupt();
                        return true;
                    }
                    sendQueue.remove(node);
                    throw e;
                }
                if (node.value.isDone) {
                    return true;
                }
                remaining = Timeouts.remaining(deadline);
                if (Timeouts.isTimeout(remaining)) {
                    sendQueue.remove(node);
                    return false;
                }
            }

        } finally {
            monitor.unlock();
        }

    }

    public Optional<E> dequeue(long timeout, TimeUnit timeUnit) throws InterruptedException {

        monitor.lock();
        try {

            // fast path
            if (sendQueue.isNotEmpty()) {
                NodeLinkedList.Node<SendRequest<E>> sender = sendQueue.pull();
                sender.value.isDone = true;
                sender.value.condition.signal();
                return Optional.of(sender.value.elementToSend);
            }
            if (Timeouts.noWait(timeout)) {
                return Optional.empty();
            }

            // wait path
            long deadline = Timeouts.start(timeout, timeUnit);
            long remaining = Timeouts.remaining(deadline);
            NodeLinkedList.Node<ReceiveRequest<E>> node =
                    receiveQueue.enqueue(new ReceiveRequest<>(monitor.newCondition()));
            while (true) {
                try {
                    node.value.condition.await(remaining, TimeUnit.MILLISECONDS);
                } catch (InterruptedException e) {
                    if (node.value.isDone) {
                        Thread.currentThread().interrupt();
                        return Optional.of(node.value.receivedElement);
                    }
                    receiveQueue.remove(node);
                    throw e;
                }
                if (node.value.isDone) {
                    return Optional.of(node.value.receivedElement);
                }
                remaining = Timeouts.remaining(deadline);
                if (Timeouts.isTimeout(remaining)) {
                    receiveQueue.remove(node);
                    return Optional.empty();
                }
            }

        } finally {
            monitor.unlock();
        }
    }
}

File: dotnet/Examples/Synchronizers/BlockingQueue.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Examples.Utils;

namespace Examples.Synchronizers
{
    public class BlockingQueue<T>
    where T: class
    {
        private class SendRequest
        {
            public bool IsDone = false;
            public readonly Condition Condition;
            public readonly T Element;

            public SendRequest(Condition condition, T element)
            {
                this.Condition = condition;
                this.Element = element;
            }
        }

        private class ReceiveRequest
        {
            public bool IsDone = false;
            public readonly Condition Condition;
            public T? Element = null;

            public ReceiveRequest(Condition condition)
            {
                this.Condition = condition;
            }
        }

        private readonly LinkedList<SendRequest> _sendRequests = new LinkedList<SendRequest>();
        private readonly LinkedList<ReceiveRequest> _receiveRequests = new LinkedList<ReceiveRequest>();
        private readonly DefaultLock _monitor = new DefaultLock();

        public bool Enqueue(T value, TimeSpan timeout)
        {
            _monitor.Lock();
            try
            {
                if (_receiveRequests.Count > 0)
                {
                    var receiver = _receiveRequests.First!;
                    _receiveRequests.RemoveFirst();
                    receiver.Value.Element = value;
                    receiver.Value.IsDone = true;
                    receiver.Value.Condition.Signal();
                    return true;
                }

                if (timeout == TimeSpan.Zero)
                {
                    return false;
                }

                var node = _sendRequests.AddLast(new SendRequest(_monitor.NewCondition(), value));
                var deadline = Deadline.FromTimeout(timeout);
                while (true)
                {
                    TimeSpan remaining = deadline.Remaining();
                    try
                    {
                        node.Value.Condition.Await(remaining);
                    }
                    catch (ThreadInterruptedException)
                    {
                        if (node.Value.IsDone)
                        {
                            Thread.CurrentThread.Interrupt();
                            return true;
                        }

                        _sendRequests.Remove(node);
                        throw;
                    }

                    if (node.Value.IsDone)
                    {
                        return true;
                    }

                    remaining = deadline.Remaining();
                    if (remaining <= TimeSpan.Zero)
                    {
                        _sendRequests.Remove(node);
                        return false;
                    }
                }
            }
            finally
            {
                _monitor.Unlock();
            }
        }

        public T? Dequeue(TimeSpan timeout)
        {
            _monitor.Lock();
            try
            {
                if (_sendRequests.Count > 0)
                {
                    var sender = _sendRequests.First!;
                    _sendRequests.RemoveFirst();
                    sender.Value.IsDone = true;
                    sender.Value.Condition.Signal();
                    return sender.Value.Element;
                }

                if (timeout == TimeSpan.Zero)
                {
                    return null;
                }

                var node = _receiveRequests.AddLast(new ReceiveRequest(_monitor.NewCondition()));
                var deadline = Deadline.FromTimeout(timeout);
                while (true)
                {
                    TimeSpan remaining = deadline.Remaining();
                    try
                    {
                        node.Value.Condition.Await(remaining);
                    }
                    catch (ThreadInterruptedException)
                    {
                        if (node.Value.IsDone)
                        {
                            Thread.CurrentThread.Interrupt();
                            return node.Value.Element;
                        }

                        _receiveRequests.Remove(node);
                        throw;
                    }

                    if (node.Value.IsDone)
                    {
                        return node.Value.Element;
                    }

                    remaining = deadline.Remaining();
                    if (remaining <= TimeSpan.Zero)
                    {
                        _receiveRequests.Remove(node);
                        return null;
                    }
                }
            }
            finally
            {
                _monitor.Unlock();
            }
        }
    }
}